pkg/provider/provider.go

```diff
@@ -308,37 +331,67 @@ func (p *Provider) GetSecretsStoreObjectContent(ctx context.Context, attrib, sec
+		if keyVaultObject.ObjectVersionHistory <= 1 {
+			file, err := p.getSecretFile(ctx, kvClient, filePermission, baseFileName, keyVaultObject, keyVaultObject.ObjectVersion, *vaultURL)
 
-		objectContent, err := getContentBytes(content, keyVaultObject.ObjectType, keyVaultObject.ObjectEncoding)
-		if err != nil {
-			return nil, err
-		}
+			if err != nil {
+				return nil, err
+			}
 
-		// objectUID is a unique identifier in the format <object type>/<object name>
-		// This is the object id the user sees in the SecretProviderClassPodStatus
-		objectUID := getObjectUID(keyVaultObject.ObjectName, keyVaultObject.ObjectType)
+			// these files will be returned to the CSI driver as part of gRPC response
+			files = append(files, file)
+			klog.V(5).InfoS("added file to the gRPC response", "file", baseFileName, "pod", klog.ObjectRef{Namespace: podNamespace, Name: podName})
+		} else {
+			// otherwise we'll generate multiple files for this object
+			secretVersions, err := p.getKeyVaultSecretVersions(ctx, kvClient, keyVaultObject, *vaultURL)
+			if err != nil {
+				return nil, err
+			}
 
-		// these files will be returned to the CSI driver as part of gRPC response
-		files = append(files, SecretFile{
-			Path:     fileName,
-			Content:  objectContent,
-			FileMode: filePermission,
-			UID:      objectUID,
-			Version:  newObjectVersion,
-		})
-		klog.V(5).InfoS("added file to the gRPC response", "file", fileName, "pod", klog.ObjectRef{Namespace: podNamespace, Name: podName})
+			for i, version := range secretVersions {
+				fileName := fmt.Sprintf("%s/%d", baseFileName, i)
+				file, err := p.getSecretFile(ctx, kvClient, filePermission, fileName, keyVaultObject, version, *vaultURL)
+
+				if err != nil {
+					return nil, err
+				}
+				files = append(files, file)
+				klog.V(5).InfoS("added file to the gRPC response", "file", fileName, "pod", klog.ObjectRef{Namespace: podNamespace, Name: podName})
+			}
+		}
 	}
 
 	return files, nil
 }
 
+func (p *Provider) getSecretFile(ctx context.Context, kvClient *kv.BaseClient, filePermission int32, fileName string, kvObject KeyVaultObject, objectVersion string, vaultURL string) (file SecretFile, err error) {
+	// fetch the object from Key Vault
+	content, newObjectVersion, err := p.getKeyVaultObjectContent(ctx, kvClient, kvObject, objectVersion, vaultURL)
+	if err != nil {
+		return SecretFile{}, err
+	}
+
+	objectContent, err := getContentBytes(content, kvObject.ObjectType, kvObject.ObjectEncoding)
+	if err != nil {
+		return SecretFile{}, err
+	}
+
+	// objectUID is a unique identifier in the format <object type>/<object name>
+	// This is the object id the user sees in the SecretProviderClassPodStatus
+	objectUID := getObjectUID(kvObject.ObjectName, kvObject.ObjectType)
+
+	return SecretFile{
+		Path:     fileName,
+		Content:  objectContent,
+		FileMode: filePermission,
+		UID:      objectUID,
+		Version:  newObjectVersion,
+	}, nil
+}
```

Comment:
Is there a reason we need a new method for just getting secret? I think we can reuse `getKeyVaultObjectContent` for multi-version scenario too. Just need to create a new `KeyVaultObject` that's passed to `getKeyVaultObjectContent`

Reply:
this was to reduce some of the duplication between the `ObjectVersionHistory <= 1` and `ObjectVersionHistory > 1` cases, but I've refactored based on your other suggestions and re-inlined this method